peakinvestigator/progress/uploader.py

```diff
@@ -0,0 +1,104 @@
+## -*- coding: utf-8 -*-
+#
+# Copyright (c) 2019, Veritomyx, Inc.
+#
+# This file is part of the Python SDK for PeakInvestigator
+# (http://veritomyx.com) and is distributed under the terms
+# of the BSD 3-Clause license.
+from peakinvestigator.progress import *
+import socket
+import ssl
+import io
+import struct
+import tarfile
+import zipfile
+
+
+class Uploader(object):
+
+    def __init__(self, host, token, progress_factory, log_level=logging.WARNING):
+        self.logger = logging.getLogger('Uploader')
+        self.logger.setLevel(log_level)
+        self.progress_factory = progress_factory
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.ssl_sock = ssl.wrap_socket(self.sock)
+        server_address = (socket.gethostbyname(host), 9999)
+        self.ssl_sock.connect(server_address)
+        self.ssl_sock.sendall(token.encode('utf-8'))
+        response = self.ssl_sock.recv(4096)
+        self.logger.info('Server response:', response)
+        if response != 'Authorized'.encode('utf-8'):
+            raise Exception('Client not authorized')
+        self.buffer = bytearray(4096)
+
+    def upload_filehandle(self, num, filetype, size, handle, progress=None):
+        self.ssl_sock.sendall(struct.pack('>I', num))
+        self.ssl_sock.sendall(struct.pack('b', filetype))
+        self.ssl_sock.sendall(struct.pack('>I', size))
+        count = handle.readinto(self.buffer)
+        while count > 0:
+            self.ssl_sock.sendall(self.buffer[:count])
+            if progress:
+                progress.update(count)
+            count = handle.readinto(self.buffer)
+        response = self.ssl_sock.recv(4096)
+        if response != 'OK'.encode('utf-8'):
+            self.close()
+            raise Exception('Problem sending file #{}: {}'.format(num, response))
+
+    def upload_file(self, filename, num, with_progress=True):
+        size = os.path.getsize(filename)
+        filetype = 0 if filename[-4:] == '.bin' else 1
+        if self.progress_factory == None:
+            with io.open(filename, 'rb') as f:
+                self.upload_filehandle(num, filetype, size, f)
+        else:
+            progress = self.progress_factory.create(total=size, unit='bytes') if with_progress else None
+            with io.open(filename, 'rb') as f:
+                self.upload_filehandle(num, filetype, size, f, progress)
+
+    def upload_files(self, filenames, start=0):
+        if self.progress_factory == None:
+            for i, filename in enumerate(filenames, start):
+                self.upload_file(filename, i, False)
+        else:
+            progress = self.progress_factory.create(total=len(filenames), unit='scans')
+            for i, filename in enumerate(filenames, start):
+                self.upload_file(filename, i, False)
+                progress.update(1)
+            progress.close()
```

Comment:
You've done this here and elsewhere. If-else statements like this are hard to read, and might indicate "code smell". Also, error prone for future changes because you have to change things in two different places.

Try this instead:
```
progress = self.progress_factory.create(total=len(filenames), unit='scans') if self.progress_factory != None else None
for i, filename in enumerate(filenames, start):
   self.upload_file(filename, i, False)
   if progress:
      progress.update(1)

if progress:
  progress.close()

```

Reply:
Branch is up to date and ready for review.

Reply:
Much better. Only see one thing, that I'll comment on.